text/0000-attribute-actions.md

```diff
@@ -0,0 +1,244 @@
+as well. While returned values from closure actions are treated as return
+values to the invocation site, classic actions use the return value to continue
+or cancel bubbling.
+
+Furthermore, even invoking an action can be inconsistent. For example to call an
+action from a component `sendAction('actionName')` is called. To call an action
+from a controller (and bubble it to the route), `send('actionName')` is used.
+
+By removing the remaining use-case for classic actions, nearly all these
+confusing discrepancies can be eliminated. Ember is easier to learn and master.
+
+# Detailed design
+
+#### Action Attachment
+
+To add an event listener to a DOM node, the author may use attributes with
+the value of an action.
+
+For example to attach a `click` event to a button, the following
+"attribute action" syntax would be used:
+
+```hbs
+<button onclick={{action 'save'}}>Save</button>
+```
+
+For comparison, to attach the `save` action with classic actions this syntax
+is used today:
+
+```hbs
+<button {{action 'save'}}>Save</button>
+```
+
+Attribute actions are more verbose in the above case, but
+have the advantage of a clear and extensible syntax. Consider attaching a
+`mouseover` event:
+
+```hbs
+<div onmouseover={{action 'save'}}>Save</div>
+```
+
+vs.
+
+```hbs
+<div {{action 'save' on='mouseOver'}}>Save</div>
+```
+
+The `onmouseover` attribute action is an evolution of the `onclick` usage.
+
+A second advantage this change is its parity with Ember component
+action passing. For example consider these components:
+
+```hbs
+{{my-button save=(action 'save')}}Save{{/my-button}}
+<my-button save={{action 'save'}}>Save</my-button>
+```
+
+This common usage of action passing is visually very similar to the new
+attribute attachment syntax. Or course an `on*` attribute passed to an
+`Ember.Component` is simply a function with no special behavior.
+
+Passing non-actions to `on*` attributes will be discouraged.
+
+* Literals should be permitted, such as `onclick="window.alert('foo')"`.
+* Non-function bindings are not permitted, and will throw from an assertion.
+  For example `onclick={{someBoundString}}` will cause an assertion to throw.
+* Non-action functions such as `onclick={{someBoundFunction}}` should cause
+  a warning to be logged. This warning will encourage the user to use
+  `onclick={{action someBoundFunction}}` or use actions as normally
+  documented.
+
+#### Handling an Attribute Action
+
+`on*` attributes will be invoked with the raw DOM event (not the jQuery event)
+as an argument. For example:
+
+```js
+// app/components/my-button.js
+export Ember.Component.extend({
+  actions: {
+    save(event) {
+      console.log('saved with click on ', event.target);
+      this.get('model').save();
+    }
+  }
+});
+```
+
+```hbs
+{{! app/templates/components/my-button.hbs }}
+<button onclick={{action 'save'}}>Save</button>
+```
+
+Actions create a closure over arguments, thus the event may not always be
+the first argument. For example:
```

Comment:
I've run into a number of occasions where the same component is used in different contexts and so its arguments end up in a different order because of the way the parent components are nested.

I then need to either come up with a new handler for each combination, or handle the various orders in the handler. Being able to use named arguments instead would be a huge win.

An example I ran into the other day (slightly simplified), we've a component which lets you change the permissions for a user in a project:

```hbs
{{change-role current=foo on-change=(action "setRole")}}
```

We currently use this in a user's profile:
```hbs
{{#each projects as |project|}}
  {{change-role current=foo on-change=(action on-change-role project)}}
{{/each}}

// setup as
{{user-profile user=user on-change-role=(action "changeRole" user)}}
```

And also in a project:
```hbs
{{#each members as |user|}}
  {{change-role current=foo on-change=(action "setRole" user)}}
{{/each}}

// setup as
{{project-admin project=project on-change-role=(action "changeRole" project)}}
```

This means `changeRole` can be called as `(project, user)` or `(user, project)`.